terraform/layer1-aws/aws-eks.tf

```diff
@@ -49,53 +49,115 @@ module "eks" {
+      }
     },
-    {
-      name                 = "ondemand"
-      instance_type        = var.eks_worker_groups.ondemand.instance_type
-      asg_desired_capacity = var.eks_worker_groups.ondemand.asg_desired_capacity
-      subnets              = module.vpc.private_subnets
-      asg_max_size         = var.eks_worker_groups.ondemand.asg_max_size
-      cpu_credits          = "unlimited"
-      kubelet_extra_args   = "--node-labels=node.kubernetes.io/lifecycle=ondemand"
-      public_ip            = false
-      additional_userdata  = file("${path.module}/templates/eks-x86-nodes-userdata.sh")
-
-      tags = local.worker_tags
+    ondemand = {
+      desired_capacity = var.node_group_ondemand.desired_capacity
+      max_capacity     = var.node_group_ondemand.max_capacity
+      min_capacity     = var.node_group_ondemand.min_capacity
+      instance_types   = var.node_group_ondemand.instance_types
+      capacity_type    = var.node_group_ondemand.capacity_type
+      subnets          = module.vpc.private_subnets
+
+      force_update_version = var.node_group_ondemand.force_update_version
+
+      k8s_labels = {
+        Environment = local.env
+        nodegroup   = "ondemand"
+      }
+      additional_tags = {
+        Name = "${local.name}-ondemand"
+      }
     },
+    ci = {
+      desired_capacity = var.node_group_ci.desired_capacity
+      max_capacity     = var.node_group_ci.max_capacity
+      min_capacity     = var.node_group_ci.min_capacity
+      instance_types   = var.node_group_ci.instance_types
+      capacity_type    = var.node_group_ci.capacity_type
+      subnets          = module.vpc.private_subnets
+
+      force_update_version = var.node_group_ci.force_update_version
+
+      k8s_labels = {
+        Environment = local.env
+        nodegroup   = "ci"
+      }
+      additional_tags = {
+        Name = "${local.name}-ci"
+      }
+      taints = [
+        {
+          key    = "nodegroup"
+          value  = "ci"
+          effect = "NO_SCHEDULE"
+      }]
+    }
+  }
+
+  worker_groups_launch_template = [
     {
-      name                    = "ci"
-      override_instance_types = var.eks_worker_groups.ci.override_instance_types
-      spot_instance_pools     = var.eks_worker_groups.ci.spot_instance_pools
-      asg_max_size            = var.eks_worker_groups.ci.asg_max_size
-      asg_min_size            = var.eks_worker_groups.ci.asg_min_size
-      asg_desired_capacity    = var.eks_worker_groups.ci.asg_desired_capacity
-      subnets                 = module.vpc.public_subnets
-      cpu_credits             = "unlimited"
-      kubelet_extra_args      = "--node-labels=node.kubernetes.io/lifecycle=spot --node-labels=purpose=ci --register-with-taints=purpose=ci:NoSchedule"
-      public_ip               = true
-      additional_userdata     = file("${path.module}/templates/eks-x86-nodes-userdata.sh")
+      name                    = "bottlerocket-spot"
+      ami_id                  = data.aws_ami.bottlerocket_ami.id
```

Comment:
Why did you decide to use different OSs (Amazon Linux 2 for AWS managed node groups and Bottlerocket for self-managed node groups)?  if it's possible it'd be better to have an option.

Reply:
EKS module doesn't support `bottlerocket` natively. If I want to use br in node_groups, I need to generate my own launch template. I tried to use asg terraform module and also tried this [approach](https://github.com/aws-samples/aws-eks-accelerator-for-terraform). But I didn't like both variants and decided to use worker_groups for bottlerocket.

You can extend your worker groups and use whatever you want)